server/src/main/java/org/elasticsearch/common/filesystem/LinuxFileSystemNatives.java

```diff
@@ -0,0 +1,230 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License
+ * 2.0 and the Server Side Public License, v 1; you may not use this file except
+ * in compliance with, at your election, the Elastic License 2.0 or the Server
+ * Side Public License, v 1.
+ */
+
+package org.elasticsearch.common.filesystem;
+
+import com.sun.jna.Native;
+import com.sun.jna.Platform;
+import com.sun.jna.Structure;
+
+import org.apache.logging.log4j.LogManager;
+import org.apache.logging.log4j.Logger;
+import org.apache.lucene.util.Constants;
+
+import java.nio.file.Files;
+import java.nio.file.Path;
+import java.time.Instant;
+import java.util.Arrays;
+import java.util.List;
+import java.util.OptionalLong;
+
+/**
+ * {@link FileSystemNatives.Provider} implementation for Linux x86-64bits
+ */
+final class LinuxFileSystemNatives implements FileSystemNatives.Provider {
+
+    private static final Logger logger = LogManager.getLogger(LinuxFileSystemNatives.class);
+
+    private static final LinuxFileSystemNatives INSTANCE = new LinuxFileSystemNatives();
+
+    /** st_blocks field indicates the number of blocks allocated to the file, 512-byte units **/
+    private static final long ST_BLOCKS_UNIT = 512L;
+
+    private final Class<? extends BaseMethod> method;
+
+    private LinuxFileSystemNatives() {
+        assert Constants.LINUX : Constants.OS_NAME;
+        assert Constants.JRE_IS_64BIT : Constants.OS_ARCH;
+
+        Class<? extends BaseMethod> method = null;
+        try {
+            logger.debug("trying to register Stat64Method...");
+            Native.register(Stat64Method.class, Platform.C_LIBRARY_NAME);
+            method = Stat64Method.class;
+        } catch (UnsatisfiedLinkError ue0) {
+            try {
+                logger.debug("trying to register XStat64Method...");
+                Native.register(XStat64Method.class, Platform.C_LIBRARY_NAME);
```

Comment:
Oh good grief this is just cursed. `stat()` in glibc is actually a macro, not a real function:

```
$ nm -AD /lib/x86_64-linux-gnu/libc.so.6 | grep 'stat$'
/lib/x86_64-linux-gnu/libc.so.6:00000000000f6d20 T __fxstat
/lib/x86_64-linux-gnu/libc.so.6:0000000000078b70 T _IO_file_stat
/lib/x86_64-linux-gnu/libc.so.6:00000000000f6d70 T __lxstat
/lib/x86_64-linux-gnu/libc.so.6:00000000001057b0 T ustat
/lib/x86_64-linux-gnu/libc.so.6:00000000000f6cd0 T __xstat
```

I think the macro might also reorganise the layout of the `struct stat` to be different from the one that `__xstat` populates. Here we're relying on the ABI of the secret `__xstat` function instead. How sure can we be that this ABI won't change?

Reply:
Libc uses symbol versioning so it would be more reliable to call `__xstat@GLIBC_2.2.5`. I don't know if specifying a symbol version is actually possible in JNA. Maybe [`FunctionMapper`](https://java-native-access.github.io/jna/4.2.1/com/sun/jna/FunctionMapper.html) can cope with the full symbol name?

Reply:
/cc @ChrisHegarty: yet more JNA-related risks ^

Reply:
@DaveCTurner Yeah, this is not ideal, but kinda unavoidable if we don't write our own C code. In general, such usage leads to me think that a small lightweight native library in ES could be worth the effort - since it would resolve many of these kinda issue ( and these issues have the potential to be significant and fail in catastrophic ways ).

Reply:
++ I reopened https://github.com/elastic/elasticsearch/issues/80903 because of this